### ISSUE TITLE
mask-rcnn detectron2 - trt engine detecting completely random objects and not accurate at all

### ISSUE THREAD

**Issue by 123mw123** [NVIDIA/TensorRT#2029]
## Description

I have converted custom detectron2 mrcnn model(5 classes) to tensorrt using these guidelines: https://github.com/NVIDIA/TensorRT/tree/main/samples/python/detectron2, but it is detecting completely random objects. Anybody else facing similar issue? 

I have made the following modifications to the code

- Modify num_classes in this line to 5. [TensorRT/create_onnx.py at main · NVIDIA/TensorRT](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2/create_onnx.py#L569)
- Set cfg.MODEL.ROI_HEADS.NUM_CLASSES = 5 and cfg.MODEL.ROI_HEADS.NMS_THRESH_TEST = 0.25 after this [line](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2/create_onnx.py#L63)
- Replace 1344 with 1024(My input image size is 1024X1024): [TensorRT/create_onnx.py at main · NVIDIA/TensorRT](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2/create_onnx.py#L170)

## Environment

**TensorRT Version**:  '8.4.0.6'
**NVIDIA GPU**:  NVIDIA TITAN RTX 
**NVIDIA Driver Version**: 470.82.01
**CUDA Version**: 11.3
**CUDNN Version**: 8.0
**Operating System**: Ubuntu 18.04
**Python Version (if applicable)**:  3.7
**PyTorch Version (if applicable)**: '1.10.0+cu113'

**Comment by nvpohanh** [comment]
@123mw123 Could you share the full scripts to repro the accuracy issue? Also, how did you train your custom detectron2 model? Asking because a lot of these accuracy issues are caused by mismatches of image preprocessing steps or grid anchor generation logic between the training phase and the inference phase. Thanks

**Comment by 123mw123** [comment]
I didn't compute accuracy with tensorrt model. I trained custom mask rcnn model using R50-FPN 3x  config provided by detectron2. I didn't modify any preprocessing steps inside detectron2. I did modify preprocessing step inside [image_batcher.py](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2/image_batcher.py#L136) to match detectron2 preprocessing like normalising image with mean pixel values. 

```
 def preprocess_image(self, image_path):
        """
        The image preprocessor loads an image from disk and prepares it as needed for batching. This includes padding,
        resizing, normalization, data type casting, and transposing.
        This Image Batcher implements one algorithm for now:
        * Resizes and pads the image to fit the input size.
        :param image_path: The path to the image on disk to load.
        :return: Two values: A numpy array holding the image sample, ready to be contacatenated into the rest of the
        batch, and the resize scale used, if any.
        """
        
        def resize_pad_cv2(image, pad_color=(0, 0, 0)):
            """
            A subroutine to implement padding and resizing. This will resize the image to fit fully within the input
            size, and pads the remaining bottom-right portions with the value provided.
            :param image: The PIL image object
            :pad_color: The RGB values to use for the padded area. Default: Black/Zeros.
            :return: Two values: The PIL image object already padded and cropped, and the resize scale used.
            """

            # Get characteristics.
            height, width = image.shape[:2]
            
            # Replicates behavior of ResizeShortestEdge augmentation.
            size = self.min_size_test * 1.0
            pre_scale = size / min(height, width)
            if height < width:
                newh, neww = size, pre_scale * width
            else:
                newh, neww = pre_scale * height, size

            # If delta between min and max dimensions is so that max sized dimension reaches self.max_size_test
            # before min dimension reaches self.min_size_test, keeping the same aspect ratio. We still need to
            # maintain the same aspect ratio and keep max dimension at self.max_size_test.
            if max(newh, neww) > self.max_size_test:
                pre_scale = self.max_size_test * 1.0 / max(newh, neww)
                newh = newh * pre_scale
                neww = neww * pre_scale
            neww = int(neww + 0.5)
            newh = int(newh + 0.5)

            # Scaling factor for normalized box coordinates scaling in post-processing. 
            scaling = max(newh/height, neww/width)

            # Padding.
            image = cv2.resize(image, (neww, newh))
            
            pad = np.ones((self.height, self.width, 3))
            pad[:,:, 0] *= pad_color[0]
            pad[:,:, 1] *= pad_color[1]
            pad[:,:, 2] *= pad_color[2]
            pad[:newh, :neww, :] = image
            return pad, scaling

        scale = None
        # image = Image.open(image_path)
        # image = image.convert(mode='RGB')

        image = cv2.imread(image_path)

        # Pad with mean values of COCO dataset, since padding is applied before actual model's
        # preprocessor steps (Sub, Div ops), we need to pad with mean values in order to reverse 
        # the effects of Sub and Div, so that padding after model's preprocessor will be with actual 0s.
        image, scale = resize_pad_cv2(image, (124, 116, 104))

        # print("resize pad", np.array(image).shape, scale)
        
        image = np.asarray(image, dtype=np.float32)
        pixel_mean = self.det2_cfg.MODEL.PIXEL_MEAN
        pixel_std = self.det2_cfg.MODEL.PIXEL_STD
        self.normalizer = lambda x: (x - pixel_mean) / pixel_std

        # Change HWC -> CHW.
        image = np.transpose(image, (2, 0, 1))
        image = (image - np.array(pixel_mean).reshape((3,1,1)))/np.array(pixel_std).reshape((3,1,1))
        # Change RGB -> BGR.

        
        return image[[2,1,0]], scale
```


Even without these modifications, the tensorrt predictions not at close to normal detectron2 predictions

**Comment by nvpohanh** [comment]
Could you share your model and the scripts/commands you used to train the model? This will help us to repro and debug the issue

**Comment by azhurkevich** [comment]
@123mw123 I've written the sample. The sample requires TensorRT >= 8.4.1 which is mentioned in the docu and it is not out yet. We've made some major updates to [PyramidROIAlign plug-in](https://github.com/NVIDIA/TensorRT/tree/main/plugin/pyramidROIAlignPlugin) that will be integrated starting 8.4.1. The release is coming very soon. 

For now you can do this to compile plug-in library separately and replace it in your TensorRT installation:
`git clone https://github.com/NVIDIA/TensorRT.git --recursive
cd TensorRT
mkdir build
cd build
cmake ..
make -j nvinfer_plugin
rm /usr/lib/x86_64-linux-gnu/libnvinfer_plugin.so*
cp -P libnv* /usr/lib/x86_64-linux-gnu/`

**Comment by 123mw123** [comment]
@nvpohanh 

Sorry for late reply. I cannot share the model. My training script looks something like this

```python3
from utils import *
from detectron2.data import DatasetCatalog, MetadataCatalog
import cv2
from detectron2.engine import DefaultTrainer
from detectron2.config import get_cfg
from detectron2.utils.visualizer import Visualizer
import random
from detectron2 import model_zoo
from detectron2.engine import DefaultPredictor
import argparse

# from detectron2.engine import default_argument_parser,default_setup

# import os
# os.environ['CUDA_LAUNCH_BLOCKING'] = "1"

# import os
# os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"
# os.environ["CUDA_VISIBLE_DEVICES"]="0,1"


if __name__ == '__main__':
    # Parse command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument('--dataset', required=True, type=str,
                        metavar="/path/to/dataset",
                        help='Full path to the dataset folder containing train and val folders')
    parser.add_argument('--chkpts', required=True, type=str,
                        metavar="/path/to/save/checkpoints",
                        help='path to save checkpoints')
    parser.add_argument('--num_iter', default=50000, type=int,
                        help='number of iterations to train')
    parser.add_argument('--batch_size', default=32, type=int,
                        help='batch size')
    parser.add_argument('--chkpt_interval', default=5000, type=int,
                        help='interval at which each chkpt should be saved')
    args = parser.parse_args()

    class_label_to_int = {"car":1,  "pedestrian":2, "traffic_light":3, "bicycle":4, "truck":5}
   


    data_dir = args.dataset
    for d in ["train", "val"]:
        DatasetCatalog.register("segment_" + d, lambda d=d: get_annotation_dicts(data_dir + d, class_label_to_int))
        MetadataCatalog.get("segment_" + d).set(thing_classes=list(class_label_to_int.keys()))
    segment_metadata = MetadataCatalog.get(data_dir)

    cfg = get_cfg()
    cfg.merge_from_file(model_zoo.get_config_file("COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml"))
    cfg.DATASETS.TRAIN = ("segment_train",)
    cfg.DATASETS.TEST = ("segment_val",)
    cfg.DATALOADER.NUM_WORKERS = 2
    cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url(
        "COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml")
    cfg.SOLVER.IMS_PER_BATCH = args.batch_size
    cfg.SOLVER.BASE_LR = 0.00025
    cfg.SOLVER.MAX_ITER = args.num_iter
    cfg.SOLVER.CHECKPOINT_PERIOD = args.chkpt_interval
    cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128
    cfg.MODEL.ROI_HEADS.NUM_CLASSES = len(class_label_to_int)
    cfg.OUTPUT_DIR = args.chkpts
    cfg.INPUT.MAX_SIZE_TRAIN = 1024
    input_sizes = [1024]
    cfg.INPUT.MIN_SIZE_TRAIN = input_sizes

    os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
    cfg.DATALOADER.FILTER_EMPTY_ANNOTATIONS = False

    trainer = DefaultTrainer(cfg)
    trainer.resume_or_load(resume=True)
    trainer.train()

```

**Comment by 123mw123** [comment]
@azhurkevich  No, I haven't tried above steps. Do you know when TensorRT >= 8.4.1 will be released?

**Comment by azhurkevich** [comment]
@123mw123 It is [live](https://docs.nvidia.com/deeplearning/tensorrt/release-notes/tensorrt-8.html#rel-8-4-1) should accessible under [TensorRT 8.4 GA](https://developer.nvidia.com/nvidia-tensorrt-8x-download). Python wheels will be available shortly.

**Comment by zhuimeng2080** [comment]
for failed to convert trt of converted.onnx, is here below:

converted.onnx
链接: https://pan.baidu.com/s/1Le5k0MjM6-yxcCVBUeNrtQ 提取码: yw2k 复制这段内容后打开百度网盘手机App，操作更方便哦
--来自百度网盘超级会员v1的分享

**Comment by nvpohanh** [comment]
@zerollzeng Could you try if you can repro the issue using this ONNX model? Thanks

**Comment by zerollzeng** [comment]
> for failed to convert trt of converted.onnx, is here below:

What does "failed to convert trt of converted.onnx" mean? it's the onnx that can reproduce the accuracy error? BTW can you upload it to Google Drive or create a Github Repo and upload it as git LFS file?

**Comment by zerollzeng** [comment]
@zhuimeng2080

**Comment by ttyio** [comment]
Closing since no activity for more than 3 weeks, please reopen if you still have question, thanks!
